src/main/java/org/mifos/connector/mpesa/camel/routes/SafaricomRoutesBuilder.java

```diff
@@ -0,0 +1,126 @@
+
+    @Override
+    public void configure() {
+
+        /*
+           Use this endpoint for receiving the callback form safaricom mpesa endpoint
+         */
+        from("rest:POST:/buygoods/callback")
+                .id("buy-goods-callback")
+                .setBody(exchange -> {
+                    String body = exchange.getIn().getBody(String.class);
+                    JSONObject object = new JSONObject(body);
+                    return object.toString();
+                });
+
+        /*
+          Rest endpoint to initiate payment for buy goods
+
+          Sample request body: {
+              "BusinessShortCode": 174379,
+              "Amount": 1,
+              "PartyA": 254708374149,
+              "PartyB": 174379,
+              "PhoneNumber": 254708374149,
+              "CallBackURL": "https://mydomain.com/path",
+              "AccountReference": "CompanyXLTD",
+              "TransactionDesc": "Payment of X"
+            }
+         */
+        from("rest:POST:/buygoods")
+                .id("buy-goods-online")
+                .process(exchange -> {
+                    String body = exchange.getIn().getBody(String.class);
+                    BuyGoodsPaymentRequestDTO buyGoodsPaymentRequestDTO = objectMapper.readValue(
+                            body, BuyGoodsPaymentRequestDTO.class);
+
+                    exchange.setProperty(BUY_GOODS_REQUEST_BODY, buyGoodsPaymentRequestDTO);
+                    logger.info(body);
+                    logger.info(buyGoodsPaymentRequestDTO.toString());
+
+                })
+                .to("direct:buy-goods-base");
+
+        /*
+         * Starts the payment flow
+         *
+         * Step1: Authenticate the user by initiating [get-access-token] flow
+         * Step2: On successful [Step1], directs to [lipana-buy-goods] flow
+         */
+        from("direct:buy-goods-base")
+                .id("buy-goods-base")
+                .log(LoggingLevel.INFO, "Starting buy goods flow")
+                .to("direct:get-access-token")
+                .process(exchange -> {
+                    exchange.setProperty(ACCESS_TOKEN, accessTokenStore.getAccessToken());
+                })
+                .log(LoggingLevel.INFO, "Got access token, moving on to API call.")
+                .to("direct:lipana-buy-goods");
+
+        /*
+         * Takes the access toke and payment request and forwards the requests to lipana API.
+         * [Password] and [TransactionType] are set in runtime and request is forwarded to lipana endpoint.
+         */
+        from("direct:lipana-buy-goods")
+                .removeHeader("*")
+                .setHeader(Exchange.HTTP_METHOD, constant("POST"))
+                .setHeader("Content-Type", constant("application/json"))
+                .setHeader("Authorization", simple("Bearer ${exchangeProperty."+ACCESS_TOKEN+"}"))
+                .setBody(exchange -> {
+                    BuyGoodsPaymentRequestDTO buyGoodsPaymentRequestDTO =
```

Comment:
Callback URL will not come from the request. It needs to be added in runtime before sending the request to Safaricom. Modify the body in direct:buy-goods-base to incorporate the callback.

Reply:
so the callbackurl need not to be passed as body, instead it should be set runtime here, Sit?

Reply:
checkout the [ChannelRequestToMpesaMapper.java](src/main/java/org/mifos/connector/mpesa/utility/ChannelRequestToMpesaMapper.java), it sets the callback URL while converting the request from entity to domain-specific DTO"

Reply:
👍